OpenSim/Common/Component.cpp

```diff
@@ -369,8 +359,26 @@ void Component::baseAddToSystem(SimTK::MultibodySystem& system) const
 
 void Component::componentsAddToSystem(SimTK::MultibodySystem& system) const
 {
-    for (const auto& compRef : _orderedSubcomponents) {
-        compRef->addToSystem(system);
+    // If _orderedSubcomponents is specified, then use this Component's
+    // specification for the order in which subcomponents are added. At a
+    // minimum the order for all immediate subcomponents must be specified.
+    if (_orderedSubcomponents.size() >= getNumImmediateSubcomponents()) {
+        for (const auto& compRef : _orderedSubcomponents) {
+            compRef->addToSystem(system);
+        }
+    }
+    else if (_orderedSubcomponents.size() == 0) {
+        // Otherwise, invoke on all immediate subcomponents in tree order
+        auto mySubcomponents = getImmediateSubcomponents();
+        for (const auto& compRef : mySubcomponents) {
+            compRef->addToSystem(system);
+        }
+    }
+    else {
+        OPENSIM_THROW_FRMOBJ(Exception, 
+            "_orderedSubcomponents specified, but does not include all "
+            "immediate subcomponents."
+        )
```

Comment:
I am fine with this implementation but the comments and this exception message are not exactly correct. If this exception is thrown, all we know is that the number of pointers in that vector is the same as the number of immediate subcomponents. I don't think a more rigorous/exact check is needed, but the comments/exception message should be weaker I think.